### PR TITLE
metrics: make meter updates lock-free

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
This is an experimental branch, please don't merge!
In the following trace, you can see that there is a lot of time spent in marking the packages that arrived on a connection in the meter. This is because we have to acquire a lock every time we want to mark a meter. This branch introduces a lock free meter that simply writes the mark into a channel and updates it periodically using a go-routine. There are some drawbacks to this: previously we started one goroutine for all meters. Now we start a goroutine and a ticker for every metric. It also means that marked changes only come available after 5 seconds, as the ticker updates the metrics in this intervals. 

I'm not sure if these drawbacks make it worth it, so please take everything with a grain of salt.

Screenshot_2020-08-14_13-28-54

### CODE REVIEW

File: metrics/meter.go [hunk 1]
```diff
@@ -161,101 +133,157 @@ func (NilMeter) Snapshot() Meter { return NilMeter{} }
...
 }
 
 // Mark records the occurrence of n events.
-func (m *StandardMeter) Mark(n int64) {
-	m.lock.Lock()
-	defer m.lock.Unlock()
-	if m.stopped {
+func (m *LockFreeMeter) Mark(n int64) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
```
Comment (holiman) [inline metrics/meter.go:237]: why is the `Rlock()` needed?

Comment (MariusVanDerWijden) [reply metrics/meter.go:237]: I thought because we read from the stopChan, but I'm dumb... We only read from it if it is closed anyway so its not a problem

File: metrics/meter_test.go [hunk 2]
```diff
@@ -16,45 +16,37 @@ func BenchmarkMeter(b *testing.B) {
...
-	defer ma.ticker.Stop()
-	m := newStandardMeter()
-	ma.meters[m] = struct{}{}
-	go ma.tick()
+	m := newLockFreeMeter()
 	m.Mark(1)
+	time.Sleep(time.Millisecond)
 	rateMean := m.RateMean()
-	time.Sleep(100 * time.Millisecond)
+	time.Sleep(5e9 + 1)
```
Comment (holiman) [inline metrics/meter_test.go:30]: Please use the time-syntax, e.g. `5 * time.Second + 1`

**Comment by holiman** [comment]
I guess the root issue is that all egress and ingress traffic uses the same two meters. It would probably be better to use a lock-free-meter only for highly-contested meters, such as those two, but leave the original meter as the default...? Othewise a lot of not-so-common meters will blow up the goroutines unnecessarily)

**Comment by holiman** [comment]
```
	ingressTrafficMeter = metrics.NewRegisteredMeter(ingressMeterName, nil)
..
	egressTrafficMeter  = metrics.NewRegisteredMeter(egressMeterName, nil)
```

**Comment by MariusVanDerWijden** [comment]
Yes, that makes sense, that why I named it `LockFreeMeter` so that I can restore `StandardMeter` easily

File: metrics/meter.go [hunk 3]
```diff
@@ -196,55 +196,45 @@ func (m *StandardMeter) Stop() {
...
+	return m.snapshot.count
 }
 
 // Mark records the occurrence of n events.
 func (m *StandardMeter) Mark(n int64) {
-	m.lock.Lock()
-	defer m.lock.Unlock()
-	if m.stopped {
+	stopped := atomic.LoadUint32(&m.stopped)
+	if stopped == 1 {
```
Comment (holiman) [inline metrics/meter.go:206]: In the 'normal path', we're going to do two separate atomic accesses (one load and one add). And in the uncommon path, we're only going to do one. 

Wild idea: how about just blindly doing the add anyway, regardless of whether it's stopped or not? Caveat: I'm assuming we don't stop/start meters, but that the stop is something that, if it does happen, happens during shutdown.

File: metrics/meter.go [hunk 4]
```diff
@@ -182,11 +184,9 @@ func newStandardMeter() *StandardMeter {
 
 // Stop stops the meter, Mark() will be a no-op if you use it after being stopped.
 func (m *StandardMeter) Stop() {
-	m.lock.Lock()
-	stopped := m.stopped
-	m.stopped = true
-	m.lock.Unlock()
-	if !stopped {
+	stopped := atomic.LoadUint32(&m.stopped)
+	atomic.StoreUint32(&m.stopped, 1)
```
Comment (holiman) [inline metrics/meter.go:188]: I think these two ops are semantically equivalent to `stopped := atomic.SwapUint32(&m.stopped, 1)`

File: p2p/peer.go [hunk 5]
```diff
@@ -303,8 +303,8 @@ func (p *Peer) handle(msg Msg) error {
 		}
 		if metrics.Enabled {
 			m := fmt.Sprintf("%s/%s/%d/%#02x", ingressMeterName, proto.Name, proto.Version, msg.Code-proto.offset)
-			metrics.GetOrRegisterMeter(m, nil).Mark(int64(msg.meterSize))
-			metrics.GetOrRegisterMeter(m+"/packets", nil).Mark(1)
+			metrics.NewRegisteredMeter(m, nil).Mark(int64(msg.meterSize))
```
Comment (MariusVanDerWijden) [inline p2p/peer.go:306]: TODO remove

File: metrics/meter.go (hunk 4 above)
Comment (MariusVanDerWijden) [reply metrics/meter.go:188]: Yep sounds good!

**Comment by MariusVanDerWijden** [comment]
Before:
```
BenchmarkMeter-16    	 6440469	       161 ns/op	       0 B/op	       0 allocs/op
```
After:
```
BenchmarkMeter-16    	163878768	         7.22 ns/op	       0 B/op	       0 allocs/op
```

File: metrics/meter.go (hunk 3 above)
Comment (MariusVanDerWijden) [reply metrics/meter.go]: I think this is not a bad idea @fjl what do you think?

File: p2p/server_test.go [hunk 6]
```diff
@@ -80,6 +81,7 @@ func startTestServer(t *testing.T, remoteKey *ecdsa.PublicKey, pf func(*Peer)) *
 		newPeerHook:  pf,
 		newTransport: func(fd net.Conn) transport { return newTestTransport(remoteKey, fd) },
 	}
+	metrics.Enabled = true
```
Comment (fjl) [inline p2p/server_test.go:84]: Please revert this change.

File: metrics/meter.go [hunk 7]
```diff
@@ -264,9 +251,19 @@ func (m *StandardMeter) updateSnapshot() {
 	snapshot.rateMean = float64(snapshot.count) / time.Since(m.startTime).Seconds()
 }
 
+func (m *StandardMeter) updateMeter() {
+	// should only run with write lock held on m.lock
+	n := atomic.LoadInt64(&m.snapshot.temp)
```
Comment (holiman) [inline metrics/meter.go:256]: We need to clear temp somewhere?
